### PR TITLE
[WIP] Define and implement new hash based groupby

### PR DESCRIPTION
**Pull request by jrhemstad**

### CODE REVIEW

File: cpp/include/cudf/groupby.hpp [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+   * ```
+   *
+   * @param requests The set of columns to aggregate and the aggregations to
+   * perform
+   * @param mr Memory resource used to allocate the returned table and columns
+   * @return Pair containing the table with each groups unique key and
+   * a set of aggregation_results for each request in the same order as was
+   * specified in `requests`.
+   */
+  std::pair<std::unique_ptr<table>, std::vector<aggregation_result>> aggregate(
```
Comment (jrhemstad) [inline cpp/include/cudf/groupby.hpp]: Ideally I'd like it if this just returned `std::vector<aggregation_result>` and the corresponding keys could be accessed via `groupby::unique_keys()`, however, I'm concerned about ensuring that the ordering of the values in the aggregation results and the unique keys are consistent if they are split into two functions.

File: cpp/src/groupby/groupby.cu [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+      _null_precedence{null_precedence} {}
+
+// Select hash vs. sort groupby implementation
+std::pair<std::unique_ptr<table>, std::vector<aggregation_result>>
+groupby::dispatch_aggregation(std::vector<aggregation_request> const& requests,
+                              cudaStream_t stream,
+                              rmm::mr::device_memory_resource* mr) {
+  // Only use hash groupby if the keys aren't sorted and all requests can be
+  // satisfied with a hash implementation
+  if (not _keys_are_sorted and
```
Comment (harrism) [inline cpp/src/groupby/groupby.cu]: Ahh, I see you already have this. Cool.

File: cpp/include/cudf/groupby.hpp [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+ * @param interpolation The desired interpolation
+ */
+std::unique_ptr<aggregation> make_quantile_aggregation(
+    std::vector<double> const& quantiles, interpolation::type interpolation);
+
+/**
+ * @brief Request for groupby aggregation(s) to perform on a column.
+ */
+struct aggregation_request {
+  column_view values;  ///< The elements to aggregate
```
Comment (harrism) [inline cpp/include/cudf/groupby.hpp]: Do the values correspond in order to the keys passed when the groupby object is created? What if you create a groupby, do some sort-based aggregations, and then try to do another sort-based aggregation with different values? How do you get these values for the request into the right order?

File: cpp/src/groupby/groupby.cu [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+                           }),
+               "Size mismatch between request values and groupby keys.");
+  if (_keys.num_rows() == 0) {
+    // TODO Return appropriate empty results
+  }
+  return dispatch_aggregation(requests, 0, mr);
+}
+}  // namespace groupby
+}  // namespace experimental
+}  // namespace cudf
```
Comment (harrism) [inline cpp/src/groupby/groupby.cu]:
```suggestion
}  // namespace cudf

```

File: cpp/src/groupby/hash/groupby.cu [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+    rmm::mr::device_memory_resource* mr) {
+  // stub
+  return std::make_pair(std::make_unique<table>(),
+                        std::vector<aggregation_result>{});
+}
+}  // namespace hash
+}  // namespace detail
+}  // namespace groupby
+}  // namespace experimental
+}  // namespace cudf
```
Comment (harrism) [inline cpp/src/groupby/hash/groupby.cu]:
```suggestion
}  // namespace cudf

```

File: cpp/src/groupby/sort/groupby.cu [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+    cudaStream_t stream, rmm::mr::device_memory_resource* mr) {
+  // stub
+  return std::make_pair(std::make_unique<table>(),
+                        std::vector<aggregation_result>{});
+}
+}  // namespace sort
+}  // namespace detail
+}  // namespace groupby
+}  // namespace experimental
+}  // namespace cudf
```
Comment (harrism) [inline cpp/src/groupby/sort/groupby.cu]:
```suggestion
}  // namespace cudf

```

File: cpp/include/cudf/groupby.hpp (hunk 1 above)
Comment (jrhemstad) [reply cpp/include/cudf/groupby.hpp]: No, but the resulting order of keys is indeterminate if a hash-based groupby is used. 

My concern is that the result of `unique_keys()` could change from one invocation to the next. For example:
```
groupby g(...);

results = g.aggregate( /* hash aggregations */); // uses hash based groupby
unique_keys = g.unique_keys(); // the order of unique_keys should match the order of results (random order)

new_result = g.aggregate( /* quantile */); // uses hash groupby
new_unique_keys = g.unique_keys(); // the order of new_unique_keys should match the order of new_result (sorted order)
// However, the order of unique_keys != the order of new_unique_keys. This seems like a problem
```

File: cpp/include/cudf/groupby.hpp (hunk 3 above)
Comment (jrhemstad) [reply cpp/include/cudf/groupby.hpp]: My assumption is that the values in a request will always be in the same order as the keys when the groupby object was constructed. The groupby object holds a `table_view` to the original keys, so it should always be able to produce the correct aggregation result.

(I'll make this explicit in the documentation).

File: cpp/include/cudf/groupby.hpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (jrhemstad) [reply cpp/include/cudf/groupby.hpp]: My current thinking is that if a sort-based aggregation is performed that an internal switch would be flipped such that all future aggregations will make use of the sorted data (if possible).

File: cpp/include/cudf/groupby.hpp [hunk 7]
```diff
@@ -0,0 +1,229 @@
...
+
+/**
+ * @brief The result(s) of an `aggregation_request`
+ */
+struct aggregation_result {
+  /// Pairs containing columns of aggregation results and their corresponding
+  /// aggregation
+  std::vector<std::pair<std::unique_ptr<column>, std::unique_ptr<aggregation>>>
+      results{};
+};
```
Comment (shwina) [inline cpp/include/cudf/groupby.hpp]: Why is the result wrapped in a struct?

Comment (jrhemstad) [reply cpp/include/cudf/groupby.hpp]: Updated doc to make it a bit clearer.

File: cpp/include/cudf/groupby.hpp (hunk 1 above)
Comment (jrhemstad) [reply cpp/include/cudf/groupby.hpp]: One idea discussed with @harrism was instead of returning a table of the unique keys from `groupby::aggregate` we could return a column of unique group IDs. This column of IDs could be used to materialize the actual unique keys that correspond to each group.

So for example:
```
[IDs, aggregation_results] = g.aggregate(...);

group_keys = g.unique_keys(IDs);
```

However, I don't really love this idea either. 

The other alternative is to require that `groupby::aggregate` always return the values in order that corresponds to the sorted order of the unique keys. This would be a unnecessary performance overhead for a hash-based implementation.

Comment (devavret) [reply cpp/include/cudf/groupby.hpp]: Option 2: (and this is made possible by the groupby object)
Don't allow the aforementioned aggregations in the `aggregate()` method. This can be done by not making an args struct for them. Then they'd need to be done using a separate `groupby().value_counts()` method call. This method can take all the args needed and can return a key table that better suits the needs of this kind of aggregation. Obvious downside is that it may need a re-sort of the value column if that column was included in a previous `aggregate()` call.

Comment (jrhemstad) [reply cpp/include/cudf/groupby.hpp]: > Take, for example, unique, nlargest (which returns less than n values if there aren't enough, for a key) and value_counts (histogram).

Ideally for all of these the returned values would be a column of lists, such that there is still only one "thing" returned per key (in this case the "thing" just happens to be a list).

File: cpp/include/cudf/groupby.hpp [hunk 8]
```diff
@@ -0,0 +1,234 @@
...
+ * `aggregation_request`.
+ *
+ * Other kinds of aggregations may derive from this class to encapsulate
+ * additional information needed to compute the aggregation.
+ */
+struct aggregation {
+  /**
+   * @brief Possible aggregation operations
+   */
+  enum Kind { SUM, MIN, MAX, COUNT, MEAN, MEDIAN, QUANTILE };
```
Comment (devavret) [inline cpp/include/cudf/groupby.hpp]: We should re-use this enum https://github.com/rapidsai/cudf/pull/3273/files#diff-19d6a8790460cbd8bb1648ee41619aeaR43-R54. Or figure out another place to keep them so that we don't have two different definitions for each aggregation kind.

Comment (devavret) [reply cpp/include/cudf/groupby.hpp]: How about this: `Kind` becomes a category. It can have categories corresponding to these: https://docs.google.com/spreadsheets/d/1PyNYv-YPVobIIg44PUPLRKscTfFgIz1Z2BPtynkSgk8/edit#gid=0&range=G:G
Then, derive from aggregation to get `reduction_aggregation` which holds a member `reduction::operator`.

Comment (jrhemstad) [reply cpp/include/cudf/groupby.hpp]: We can't reuse the enum because there are enums on that list that are not supported by groupby.

Comment (jrhemstad) [reply cpp/include/cudf/groupby.hpp]: I don't know what you mean by "category".

Comment (devavret) [reply cpp/include/cudf/groupby.hpp]: > We can't reuse the enum because there are enums on that list that are not supported by groupby.

But all of those are eventually going to be supported for groupby right? There can be another way to indicate that even though it exists in the enum, it's not currently supported for groupby, but will be in the future.

And by category, I mean:
```c++
enum Kind { REDUCTION, SCAN, COMPACTION ... };
```

**Comment by jrhemstad** [comment]
So I'm torn on what the right interface is for `groupby::aggregate`. 

The fundamental question to answer is "How does the caller of `groupby::aggregate` know what order the resulting values are in and what their group labels are?" 

Here are the options I can think of:

1. `groupby::aggregate` always returns `pair<table, vector<results>` with the table of keys and the aggregation results. [Example. ](https://github.com/rapidsai/cudf/blob/4782d5a03d85975c5a455d8478bbd576553e7d18/cpp/include/cudf/groupby.hpp#L191)
   - Pros: 
      - No sort required
   - Cons
      - Complex API (returns both a table and the aggregation results)
      - Requires materializing the output keys table _every_ time.
      
2. `groupby::aggregate` only returns `vector<results>` guaranteed to be in the same sorted order as `groupby::groups`. [Example](https://github.com/rapidsai/cudf/blob/776a6ac971e6b0aa4602bf7ab1448b9d1bf11cd8/cpp/include/cudf/groupby.hpp#L190).
   - Pros:
      - Cleanest API as `groupby::aggregate` only returns aggregation results and nothing else
   - Cons:
      - Requires unnecessary sorting of results which introduces overhead
      - The sorted order may not be the sorted order the user wants, which requires them to do _another_ sort
3. `groupby::aggregate` returns _something_ (like the opaque object) that you pass into `groupby::groups` to materialize the keys. [Example.](https://github.com/rapidsai/cudf/blob/41e3a5b09c9d15c86fb739b3aecb99082421a4bd/cpp/include/cudf/groupby.hpp#L209)
   - Pros:
      - No sort required
      - Opportunity to optimize by avoiding _always_ materializing output keys (when and if they keys are materialized is hidden in the opaque `group_labels`
   - Cons
     - Complex API (returns both a `group_labels` and aggregation_results)
     - Requires passing `group_labels` from `groupby::aggregate` to `groupby::groups`

**Comment by harrism** [comment]
What if aggregate returned just an opaque object `aggregation`, which has a method to access `results()`, a method to access the `group_labels()`, and possibly even another method to access a pair of `<labels, results>`. The latter would be unsorted, but provide the direct correspondence if the user prefers that. The former would sort only once it is called.

**Comment by jrhemstad** [comment]
> What if aggregate returned just an opaque object `aggregation`, which has a method to access `results()`, a method to access the `group_labels()`, and possibly even another method to access a pair of `<labels, results>`. The latter would be unsorted, but provide the direct correspondence if the user prefers that. The former would sort only once it is called.

What would `group_labels()` return? It would probably have to be a `shared_ptr<table>` so it can be shared by the various aggregation results. But if we start returning a `shared_ptr<table>` then we get into some uncharted areas of ownership problems.

**Comment by devavret** [comment]
About option 3, I don't foresee many cases where one **wouldn't** need a mapping between what aggregate values correspond to what keys. Except maybe the case where you looking for a reduction of a groupby reduction like what's the maximum groupwise count, or what's the minimum standard deviation etc. In almost all scenarios, one would have to make a second call to `groupby::groups` so the performance advantages don't seem too much. 

This, coupled with the fact that the API requires you to pass around an object between methods of the _same_ class makes it my least favourite.

In general I'd expect the group order to be fixed between calls to the same groupby object. That means option 2 is my most preferred. There's 3 cases in this:
1. How to ensure the order between two hash based calls are the same
2. How to ensure the order stays same after one hash based call and a subsequent sort based call
3. The same thing but when a sort based precedes a hash based call.

**For 1,**
Generate the hash table on the first call. The order after one insertion is now fixed. Re-use this hash table in all subsequent calls. Since you don't check whether it was a new insertion or a repeat insertion in `build_aggregation_map` kernel, it shouldn't matter if it was populated from last time. https://github.com/rapidsai/cudf/blob/16de9218d6599edec38157f6ffc640d17e493b16/cpp/src/groupby/hash/legacy/groupby_kernels.cuh#L102-L106 (There's no use of `result.second`). Of course, the `sparse_output_values` table will be rebuilt in each call.

**For 2,**
In sort-based groupby, it doesn't actually matter whether the keys are sorted (Unless the user specifies the key sort order in `groupby` object's constructor). What matters is that the values are sorted within the group. This is what's required to enable sort-dependent aggregations like `quantile`, `nth` etc. If a hash-based aggregation happens first, it can be arranged to store that key order in the `sort_helper` object so that it becomes the defacto order from now on.

**For 3,**
If a sort-based aggregation happens first, then we can get the unique key labels from the `sort_helper` and create a hash map.

----------------------------------
We don't have to do all this right now. This can be an optimization for future. 

Option 2 seems like the best API in terms of correctness and usability. The only cons are performance related.
>The sorted order may not be the sorted order the user wants, which requires them to do another sort

This is not really an issue because the user should specify the desired sort order in the `groupby()` constructor. Plus, it's not like re-sorting isn't an issue with the other two. This con applies to all three.
That leaves us with only one con: requirement to sort the result. And what i demonstrated above is that in the future we can take care of that as well. Right now, we sort the result of the hash based aggs. It's still faster than sort based coz size(unique_keys) < size(all_keys)

### So here's my suggestion:
- Go with option 2.
- Sort the hashed aggs' result for now.
- We do the aforementioned optimizations later

**Comment by jrhemstad** [comment]
>  I don't foresee many cases where one wouldn't need a mapping between what aggregate values correspond to what keys
> In almost all scenarios, one would have to make a second call to `groupby::groups` so the performance advantages don't seem too much.

I agree that 100% of the time a user will do:
```
[labels, values] = groupby::aggregate(...);
keys = groupby::groups(labels);
```
The goal is not to improve performance by avoiding `groupby::groups`, instead, the optimization opportunity is in how much work is actually done in `groupby::groups`. 

For example,
```
[labels,values] = groupby::agreggate(...); // sort-based
keys = groupby::groups(labels); // materializes the unique keys

[labels2, values2] = groupby::aggregate(...); // sort-based again
keys2 = groupby::groupbys(labels2); // Here is the optimization opportunity. _keys2_ are the same as _keys_, so they don't actually need to be re-materialized 
```

> Re-use this hash table in all subsequent calls.

This is an interesting idea I hadn't considered. **However**, even if you re-use the same hash map, there's currently no guarantee that when you extract results from the map that the output order will be the same, i.e., if you were to extract the same results from the same hash map, there's no guarantee that the order will be the same. That said, we may be able to come up with a way to extract results from a hash map in a repeatable order. 

> If a hash-based aggregation happens first, it can be arranged to store that key order in the sort_helper object so that it becomes the defacto order from now on.

This sounds like just trading the overhead from the hash-based path to the sort-based path. Either way, you have to do extra work you wouldn't otherwise have to do, i.e., either sort the results in the hash-based path, _or_ rearrange the key order in the sort-path to match the hash-path. 

> For 3,
If a sort-based aggregation happens first, then we can get the unique key labels from the sort_helper and create a hash map.

I'm not following how this would work. Even if you know the unique keys ahead of time and insert them into a hash map, you have no idea what order the keys will end up in the map. Furthermore, see my above comment about the indeterminism in the order results are extracted from a hash map. Finally, this would also be extraneous overhead as it would require you to first build the hash map with just the unique keys, and then "build" it again when doing the aggregations.  

> This is not really an issue because the user should specify the desired sort order in the groupby() constructor. 

The user is not specifying the _desired_ sort order in the constructor, they are specifying whether or not it is _already_ sorted (in order to avoid having to sort it again to perform a sort groupby). I am resistant to complicating the `groupby` object by giving the option to request the results be sorted (i.e., in a call to `aggregate`). It is trivial for the user to sort the results themselves. 

**In Summary**

If you think about it at a fundamental level, a hash-based groupby will always produce results in a different order than a sort-based groupby. If you want them to be in the same order, it requires you to pay a performance penalty and do extra work to get them in the same order (either sorting the groupby result or rearranging the sort-based result to match the hash result). 

Solutions 1 & 3 are the only two solutions that avoid any extra work to force the order of results to be the same between a sort vs hash groupby.

**Comment by abellina** [comment]
In Spark's use case, the grouping key is required for a group by. For reduction aggregates, the grouping key doesn't matter. This PR is not merging the reduction interface into the group by, as far as I can tell, is that the plan? 

If the plan is to merge the reduction operation into this also, then having a lazily evaluated grouping key table could make sense, though in that case that table is tiny (1 row). Is producing a 1 row table eagerly still a big cost?

Adding extra sorting is going to make the hash aggregate slower. @jrhemstad Have you run some quick benchmarks on what the impact might be? Just curious. I like @harrism's proposal to have `aggregate` return an opaque object, but I am trying to understand the ownership problem @jrhemstad pointer at:

> What would group_labels() return? It would probably have to be a shared_ptr<table> so it can be shared by the various aggregation results. But if we start returning a shared_ptr<table> then we get into some uncharted areas of ownership problems.

Does it mean that `aggregation_result` will have a `table`, could you elaborate @jrhemstad?

**Comment by jrhemstad** [comment]
> In Spark's use case, the grouping key is required for a group by. For reduction aggregates, the grouping key doesn't matter. This PR is not merging the reduction interface into the group by, as far as I can tell, **is that the plan?**

No, definitely not. This was just misunderstanding.

> I am trying to understand the ownership problem

So let's say we did the following:

```c++
class aggregation_result{
   std::shared_ptr<table> keys;
   std::vector< unique_ptr<column> > values;
};

std::vector<aggregation_result> groupby::aggregate(...);
```

The `keys` table is the same for all `aggregation_result`s, and so in order to avoid having to materialize a unique table for every result, it should be shared among them via a `shared_ptr`. 

However, the `groupby` object may _also_ hold on to a reference to that very same `keys` table because it may want to return it _again_ to satisfy a subsequent `aggregate` call. 

Here's the problem, in Cython, I suspect they'll want to do the following:
```c++
results = groupby::aggregate(...);

// this will _release_ all of the constituent columns, effectively destroying
// the `keys` table, which invalidates the reference that the `groupby` object is  holding on to
cython_table = from_cpp_table(results.keys); 

```
@shwina can confirm. 

Here's an alternative idea:

```c++
class aggregation_result{
   table_view keys;
   std::vector< unique_ptr<column> > values;
};

std::vector<aggregation_result> groupby::aggregate(...);
```

Now the returned keys is a `table_view`. The `groupby` object owns the table that is viewed by the `table_view` in the `aggregation_result`. This means that as soon as the `groupby` object is destroyed, that `keys` view is invalid. However, the caller could make a copy:

```c++
groupby g(...);
results = g.aggregate(...);
table_view keys_view = results.keys;

table keys_copy(keys_view); // this copies the keys from the groupby object

g.~groupby(); // groupby object is destroyed

// keys_view is no longer valid
// keys_copy is valid
```
The problem with this solution is it incurs unnecessary copies.

**Comment by abellina** [comment]
@jrhemstad from this https://github.com/rapidsai/cudf/blob/41e3a5b09c9d15c86fb739b3aecb99082421a4bd/cpp/src/groupby/groupby.cu#L67, the `groupby` constructor takes a `table_view` for `keys`. That said I think this `keys` is different than the one we are describing (perhaps we should distinguish the two with a different name).

On `aggregation_result` returning a `table_view` instead of a shared pointer of `table` seems like a safe choice if mutating it would affect `groupby`. The other option is to not have this state in `groupby` at all... is running multiple `aggregation_result`s from the same `groupby` important? I am not sure: `aggregate` already takes a vector of `aggregation_request`. What's the use case here?

**Comment by jrhemstad** [comment]
> That said I think this `keys` is different than the one we are describing (perhaps we should distinguish the two with a different name).

Yeah, my example was ambiguous. I renamed to `unique_keys` to disambiguate. 

> The other option is to not have this state in groupby at all... is running multiple aggregation_results from the same groupby important?

Maybe? 

What seems more likely is something like this:
```
groupby g(...);

// Get the size of each group (you asked for this one in a previous issue)
auto group_sizes = g.group_sizes();

// user wants to know the unique keys
auto unique_keys = g.unique_keys(); // or g.groups(), naming doesn't matter

// now perform an actual aggregation
auto aggregation_results = g.aggregate(...);
```

Ideally we'd like to save state inside of `g` in order to eliminate redundant computations in satisfying the calls to `group_sizes(), unique_keys(), aggregate(...)`. Part of that state would probably be the table of `unique_keys`.

**Comment by abellina** [comment]
> // Get the size of each group (you asked for this one in a previous issue)

Ah, so it's all my fault :)

So in this issue: https://github.com/rapidsai/cudf/issues/3131, I didn't add a group by, and that was unclear. I can make that change there. In the group by case, I expect a row per grouping key with the count not skipping nulls as the value column.

Could these be other types of `aggregation_request`, since the result exactly matches the shape of the grouping key?

```
// Get the size of each group (you asked for this one in a previous issue)
auto group_sizes = g.group_sizes();

// user wants to know the unique keys
auto unique_keys = g.unique_keys(); // or g.groups(), naming doesn't matter
```

**Comment by jrhemstad** [comment]
> Could these be other types of `aggregation_request`, since the result exactly matches the shape of the grouping key?

To be clear, your suggestion was to do away with methods like `groupby::group_sizes` and `groupby::unique_keys` and only have `groupby::aggregate`. We could then provide special `aggregation_requests` that do the same thing as `groupby::group_sizes/unique_keys`.

I believe this is inspired by a Spark workflow where you know everything you're going to do with a "groupby" ahead of time.

In contrast, with Pandas, a user may create a groupby object and do all sorts of things with it. We'd like to eliminated as much of that redundant work as possible.

**Comment by jrhemstad** [comment]
Talking with @kkraus14 and @shwina, Python prefers Option 1.

For every call to `aggregate`, they expect a newly allocated `table` of the unique keys. 

This still allows us an opportunity to optimize what we return by storing this table internally and just returning a copy. However, there does not appear to be a use case for returning anything other than `table`, (i.e., returning a `shared_ptr<table>` or `table_view`). 

As such, I'm 99% leaning towards Option 1 unless anyone has strong feelings against this.

**Comment by jrhemstad** [comment]
@devavret @harrism I've pushed what I believe to be the final API design.

- I separated out the `aggregation` class into a separate `aggregation.hpp` header. This can be reused by other things like reductions, rolling, etc. We'll just have to be explicit about what operations are supported. I'm still on the fence about this, but I think I'm out numbered.

- I went with Option 1, `groupby::aggregate` always returns keys/values. Neither the Spark nor Python team had any opportunity to avoid materializing the unique keys table, so this obviates any need for Option 3. By linking them in the `aggregate` call avoids any of the gymantics that would be required for Option 2. (Notice in the documentation I explicitly say the ordering is arbitrary and successive calls can return different orders).

File: cpp/include/cudf/aggregation.hpp [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+/**
+ * @brief Factory to create a QUANTILE aggregation
+ *
+ * @param quantiles The desired quantiles
+ * @param interpolation The desired interpolation
+ */
+std::unique_ptr<aggregation> make_quantile_aggregation(
+    std::vector<double> const& q, interpolation i);
+}  // namespace experimental
+}  // namespace cudf
```
Comment (harrism) [inline cpp/include/cudf/aggregation.hpp]:
```suggestion
}  // namespace cudf

```

File: cpp/include/cudf/detail/aggregation.hpp [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+  quantile_aggregation(std::vector<double> const& q,
+                       experimental::interpolation i)
+      : aggregation{QUANTILE}, _quantiles{q}, _interpolation{i} {}
+  std::vector<double> _quantiles;              ///< Desired quantile(s)
+  experimental::interpolation _interpolation;  ///< Desired interpolation
+};
+
+}  // namespace detail
+}  // namespace experimental
+}  // namespace cudf
```
Comment (harrism) [inline cpp/include/cudf/detail/aggregation.hpp]:
```suggestion
}  // namespace cudf

```

File: cpp/include/cudf/detail/aggregation.hpp [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+ *
+ * Other kinds of aggregations may derive from this class to encapsulate
+ * additional information needed to compute the aggregation.
+ */
+class aggregation {
+ public:
+  /**
+   * @brief Possible aggregation operations
+   */
+  enum Kind { SUM, MIN, MAX, COUNT, MEAN, MEDIAN, QUANTILE };
```
Comment (harrism) [inline cpp/include/cudf/detail/aggregation.hpp]: Why is this class in a `detail` file? Seems to me that the type of the aggregations is really part of the main user API, so should be in the non-detail header. Evidence for this is that it's not in the detail namespace.

File: cpp/include/cudf/aggregation.hpp [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+
+namespace cudf {
+namespace experimental {
+/**
+ * @brief Base class for abstract representation of an aggregation.
+ */
+class aggregation;
+
+/// Factory to create a SUM aggregation
+std::unique_ptr<aggregation> make_sum_aggregation();
```
Comment (harrism) [inline cpp/include/cudf/aggregation.hpp]: Why do we need a factory for every trivial aggregation type? Seems to me the reason for this complexity is due to one outlier: quantiles. Why not treat quantiles specially rather than making everything more complex? Just use an enum class for all the simple aggregations, and complex aggregations can be separate. (?)

File: cpp/src/aggregation/aggregation.cpp [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+  return std::make_unique<aggregation>(aggregation::MEDIAN);
+}
+/// Factory to create a QUANTILE aggregation
+std::unique_ptr<aggregation> make_quantile_aggregation(
+    std::vector<double> const& q, interpolation i) {
+  aggregation* a = new detail::quantile_aggregation{q, i};
+  return std::unique_ptr<aggregation>(a);
+}
+}  // namespace experimental
+}  // namespace cudf
```
Comment (harrism) [inline cpp/src/aggregation/aggregation.cpp]:
```suggestion
}  // namespace cudf

```

**Comment by harrism** [comment]
What's the plan with this PR? Will it be open until the API is implemented, or do you want to merge the new API to allow simultaneous work?

File: cpp/include/cudf/detail/aggregation.hpp (hunk 11 above)
Comment (jrhemstad) [reply cpp/include/cudf/detail/aggregation.hpp]: It's an opaque(ish) type. There's no reason for users to know about the internals of the class, so it's only forward declared in the public headers. The detail header contains the definition.

File: cpp/include/cudf/aggregation.hpp (hunk 12 above)
Comment (jrhemstad) [reply cpp/include/cudf/aggregation.hpp]: > Seems to me the reason for this complexity is due to one outlier: quantiles.

For _now_, sure, but as we add more complex aggregations, factories with arguments will become the norm rather than the outlier. For example, Spark wants a "count" aggregation where you can parameterize whether or not null values are counted.

> Just use an enum class for all the simple aggregations, and complex aggregations can be separate.

All aggregations need to share a common base class in order to store all possible aggregations in a container like `std::vector<std::unique_ptr<aggregation>>`. The _derived_ classes are separate, but they all need a common representation.

**Comment by jrhemstad** [comment]
> What's the plan with this PR? Will it be open until the API is implemented, or do you want to merge the new API to allow simultaneous work?

It should be merged so that @devavret and I can proceed to implement against the API in parallel.

File: cpp/include/cudf/aggregation.hpp (hunk 12 above)
Comment (harrism) [reply cpp/include/cudf/aggregation.hpp]: But do we need factories for every one, or would a constructor suffice?

Comment (jrhemstad) [reply cpp/include/cudf/aggregation.hpp]: What would the constructor look like? Probably something like:
```
aggregation(aggregation::Kind kind);
```

but then you'd have to document and error check for all the values of `kind` that are acceptable (e.g., only `min/max/sum`,etc). Plus `aggregation` could no longer be an opaque type. 

Furthermore, it has to be a `unique_ptr`, so users would have to do something like:
```
make_unique<aggregation>(...);
```

It's not really any more code to just use a factory, and the factory is explicit and unambiguous.

**Comment by jrhemstad** [comment]
rerun tests

File: cpp/include/cudf/aggregation.hpp [hunk 14]
```diff
@@ -0,0 +1,68 @@
...
+/// Factory to create a MEDIAN aggregation
+std::unique_ptr<aggregation> make_median_aggregation();
+
+/**
+ * @brief Factory to create a QUANTILE aggregation
+ *
+ * @param quantiles The desired quantiles
+ * @param interpolation The desired interpolation
+ */
+std::unique_ptr<aggregation> make_quantile_aggregation(
```
Comment (devavret) [inline cpp/include/cudf/aggregation.hpp]: Should likely rename to `make_quantile_aggregation_args` or `make_quantile_agg_args` or `make_quantile_args`. Basically something that means it's the args.

Comment (jrhemstad) [reply cpp/include/cudf/aggregation.hpp]: I'd argue the factory is not making the "args", it's making the aggregation itself. 

A `quantile` aggregation is inseparable from its arguments. It's meaningless to request a `quantile` without specifying _which_ quantiles you want and how to interpolate them. 

Likewise with standard deviation, we'd do:
```
unique_ptr<aggregation> make_stddev_aggregation(int ddof);
```

not 
```
unique_ptr<aggregation> make_stddev_aggregation_args(int ddof);
```

Plus, then the naming would be inconsistent with the factories that don't take arguments.

File: cpp/include/cudf/aggregation.hpp (hunk 12 above)
Comment (harrism) [reply cpp/include/cudf/aggregation.hpp]: This is why I asked. I didn't know the things you describe. Thanks for explaining so patiently. :-)
